src/sentry/services/hybrid_cloud/sig.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+import inspect
+import itertools
+from typing import Any, Callable, Sequence, Tuple, Type
+
+import pydantic
+
+from sentry.services.hybrid_cloud import ArgumentDict
+
+
+class SignatureSerializationException(Exception):
+    pass
+
+
+class SerializableFunctionSignature:
+    """Represent a function's parameters and return type for serialization."""
+
+    def __init__(self, base_function: Callable[..., Any], is_instance_method: bool = False) -> None:
+        super().__init__()
+        self.base_function = base_function
+        self.is_instance_method = is_instance_method
+
+        self._parameter_model = self._create_parameter_model()
+        self._return_model = self._create_return_model()
+
+    def get_name_segments(self) -> Sequence[str]:
+        return (self.base_function.__name__,)
+
+    def _setup_exception(self, message: str) -> SignatureSerializationException:
+        name = ".".join(self.get_name_segments())
+        return SignatureSerializationException(f"{name}: {message}")
+
+    def _validate_type_token(self, token: Any) -> None:
+        """Check whether a type token is usable.
+
+        Strings as type annotations, which Mypy can use if their types are imported
+        in an `if TYPE_CHECKING` block, can't be used for (de)serialization. Raise an
+        exception if the given token is one of these.
+
+        We can check only on a best-effort basis. String tokens may still be nested
+        in type parameters (e.g., `Optional["RpcThing"]`), which this won't catch.
+        Such a state would cause an exception when we attempt to use the signature
+        object to (de)serialize something.
+        """
+        if isinstance(token, str):
+            name = ".".join(self.get_name_segments())
+            raise ValueError(f"Type annotations on {name} must be actual type tokens, not strings")
```

Comment:
Seems like you could use `_setup_exception()` here and remove a layer of try/except on 59-64. That would change the exception types in a few other methods though. The other call sites are all related to signature serialization as well.

Reply:
Good call. Looks like that was originally [done](https://github.com/getsentry/sentry/commit/80c5f777a2986bebfa0190920b90ad2fbf3cff5e) so that the extra try-except layer can add the parameter name to the message. https://github.com/getsentry/sentry/pull/58529/commits/3e6e4c03af0779831a60d8d3ef96fdf95440b43e lets us do the same.